File: src/mysql/targets/antlr4wasm/SimpleErrorListener.ts
/*
 * Copyright (c) Mike Lischke. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */

import {
    ANTLRErrorListener as AEL, type ATNConfigSet, type ATNSimulator, type BitSet, type Parser,
    type RecognitionException, type Recognizer, type Token
} from "./antlr4-runtime";

import type { IParserErrorInfo } from "../misc/types";

const ANTLRErrorListener = AEL.extend<AEL>("AEL", {});
type ANTLRErrorListener = InstanceType<typeof ANTLRErrorListener>;

export class SimpleErrorListener extends ANTLRErrorListener {
    public constructor(private errors: IParserErrorInfo[]) {
        super();
    }

    public override syntaxError(recognizer: Recognizer<ATNSimulator>, offendingSymbol: Token | null, line: number,
        charPositionInLine: number, message: string, e: RecognitionException | null): void {

        if (length === 0) {
            length = 1;
        }

        this.errors.push({
            message,
            tokenType: offendingSymbol?.getType() ?? 0,
            charOffset: offendingSymbol?.getStartIndex() ?? 0,
            line,
            offset: charPositionInLine,
            length
        });
    };

    public override reportAmbiguity(recognizer: Parser, /*dfa: DFA,*/ startIndex: number, stopIndex: number, exact: boolean,
        ambigAlts: BitSet, configs: ATNConfigSet): void { }
    public override reportAttemptingFullContext(recognizer: Parser, /*dfa: DFA,*/ startIndex: number, stopIndex: number,
        conflictingAlts: BitSet, configs: ATNConfigSet): void { }
    public override reportContextSensitivity(recognizer: Parser, /*dfa: DFA,*/ startIndex: number, stopIndex: number,
        prediction: number, configs: ATNConfigSet): void { }

}

File: src/mysql/targets/antlr4-java/misc/SimpleErrorListener.java
package misc;

import org.antlr.v4.runtime.*;
import org.antlr.v4.runtime.atn.ATNConfigSet;
import org.antlr.v4.runtime.dfa.DFA;

import java.util.BitSet;
import java.util.List;

public class SimpleErrorListener implements ANTLRErrorListener {

    private List<ParserErrorInfo> errors;

    public SimpleErrorListener(List<ParserErrorInfo> errors) {
        this.errors = errors;
    }

    @Override
    public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int column, String message, RecognitionException e) {
        Token token = offendingSymbol instanceof Token ? (Token) offendingSymbol : null;
        int tokenType = token == null ? 0 : token.getType();
        int charOffset = token == null ? 0 : token.getStartIndex();
        int length = token == null ? 0 : token.getStopIndex() - token.getStartIndex() + 1;

        this.errors.add(new ParserErrorInfo(message, tokenType, charOffset, line, column, length));
    }

    @Override
    public void reportAmbiguity(Parser parser, DFA dfa, int i, int i1, boolean b, BitSet bitSet, ATNConfigSet atnConfigSet) {
    }

    @Override
    public void reportAttemptingFullContext(Parser parser, DFA dfa, int i, int i1, BitSet bitSet, ATNConfigSet atnConfigSet) {
    }

    @Override
    public void reportContextSensitivity(Parser parser, DFA dfa, int i, int i1, int i2, ATNConfigSet atnConfigSet) {
    }
}
